server/src/main/java/org/opensearch/search/approximate/ApproximateScoreQuery.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.approximate;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+
+/**
+ * Base class for a query that can be approximated.
+ *
+ * This class is heavily inspired by {@link org.apache.lucene.search.IndexOrDocValuesQuery}. It acts as a wrapper that consumer two queries, a regular query and an approximate version of the same. By default, it executes the regular query and returns {@link Weight#scorer} for the original query. At run-time, depending on certain constraints, we can re-write the {@code Weight} to use the approximate weight instead.
+ */
+public class ApproximateScoreQuery extends Query {
+
+    private final Query originalQuery;
+    private final ApproximateableQuery approximationQuery;
+
+    private SearchContext context;
```

Comment:
I suggest that instead of holding the `SearchContext`, the `setContext(...)` method should call `approximationQuery.canApproximate()` and set a mutable `Query resolvedQuery` field (to point to the right query).

Then we can enforce some "SetOnce" style semantics:
1. If someone calls `setContext` and `resolvedQuery` is not null, throw an exception.
2. If someone calls `createWeight` and `resolvedQuery` is null, throw an exception.